rat/compiler.py

```diff
@@ -329,15 +332,24 @@ def walk_Statement(self, node: ast.Statement):
         self.walk(node.right)
 
         traces = defaultdict(lambda: [])
-        for row in primary_variable.itertuples():
+        for row_number, row in enumerate(primary_variable.itertuples()):
             executor = OpportunisticExecutor(self.variable_table, row._asdict(), self.trace_by_reference)
             executor.walk(node)
 
             for traced_node, value in executor.values.items():
-                traces[traced_node].append(value)
+                traces[traced_node].append((row_number, value))
+
+        dense_size = len(primary_variable)
+        for traced_node, trace in traces.items():
+            idxs, values = zip(*trace)
+
+            numpy_idxs = numpy.array(idxs)
+            numpy_values = numpy.array(values)
+
+            dense_values = numpy.zeros(dense_size, dtype=numpy_values.dtype)
+            dense_values[numpy_idxs] = numpy_values
```

Comment:
This change makes it so that any time we trace something, we pad stuff out so we have arrays with length equal to the size of the domain of the primary variable.

Ifelses make it possible for variables get used only some of the time during tracing. vmapping requires that all input arrays be equal length. This is padding everything out to avoid vmap errors.

Reply:
Hmm. Can there be a problem with having zeros where we shouldn't have them? What if we had something like `ifelse(x < 1e-5, 0, log(x))`?